### ISSUE TITLE
Add support for ExecuteResultAsync 

### ISSUE THREAD

**Issue by thuannguy** [aspnet/AspNetWebStack#113]
The Asp.Net core counterpart has support for [ExecuteResultAsync](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.iactionresult.executeresultasync?view=aspnetcore-2.0) which is very useful. Without that, it is very hard to call a method asynchronously from ExecuteResult. Could you please consider adding support for that?

**Comment by abatishchev** [comment]
So does [`IHttpActionResult.ExecuteAsync()`](https://msdn.microsoft.com/en-us/library/system.web.http.ihttpactionresult.executeasync(v=vs.118).aspx#M:System.Web.Http.IHttpActionResult.ExecuteAsync(System.Threading.CancellationToken)). Or you're looking for something different?

**Comment by thuannguy** [comment]
Please let me clarify the request. I meant ActionResult of Asp.Net MVC as in the simplified example below:

    public class MyController : Controller
    { 
        public ActionResult Resolve()
        {
            // Simplified code
            return new MyActionResult(...);    // the ExecuteResult method of this action result will be invoked by MVC itself.
        }
    }

Therefore, IHttpActionResult.ExecuteAsync is not the one I need 😃

**Comment by jdaigle** [comment]
If I were to create a PR adding support for async MVC ActionFilters and async MVC ActionResults, how likely is Microsoft to seriously consider it? I feel this could be a big win for users that can’t upgrade to Core (i.e. having to support a hybrid WebForms app).

I already created a prototype several months ago, including refactoring much of the Begin/End async code internal to the Controller execution pipeline - it’s not difficult to add support, but doing so without introducing breaking changes is the hard part.

**Comment by thuannguy** [comment]
@jdaigle yep, I can't upgrade to Core because many libs I need aren't available there yet and AFAICT it will take .Net/Asp.Net/Azure team a long time to support them. Therefore, I have a (slim) hope that supporting async ExecuteResult would be a less resistant path 😛

**Comment by mkArtakMSFT** [comment]
@thuannguy, which libraries are missing, that block you from switching to ASP.NET Core instead?
@jdaigle, unfortunately we do not plan to add/accept this feature. Rather, we'd be happy to enable missing APIs in ASP.NET Core instead to enable the switch.

**Comment by jdaigle** [comment]
@mkArtakMSFT That's sad to hear, but not surprising given the scope of the changes. And unfortunately it's not really "missing APIs" so much as having a giant hybrid WebForms and MVC5 application that we'll need to support and add new features to for some time to come.

We might have considered "forking" MVC5 and using a private build to get the features we want. Except that strong-naming prevents us from doing that (and I suppose that you're unlikely to release the signing key for all to use).

But it might be possible to reimplement most of the MVC pipeline in a separate project, since most of the pipeline is implemented in `Controller`. We just need to create our own implementation of `IAsyncController`. So much the async implementation could be significant cleaned up and simplified using `System.Threading.Tasks.Task`, rather than the Begin/End asynchronous programming model.

**Comment by thuannguy** [comment]
@mkArtakMSFT 
Some of them are 
https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/852
and https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/476. I can understand that they have limited resources and thus the WSTrust request won't be ready anytime soon.

I have the same "giant hybrid WebForms and MVC5 application" situation which is another adversary. The ability to do true async controller now would ease my pain a lot before I can plan for a full upgrade to Core.

**Comment by mkArtakMSFT** [comment]
@jdaigle, it is possible to get the same identity on your builds as you'd get from Microsoft provided builds using so-called [Public Signing](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/public-signing.md). Hope this helps.
